File: mobile/lib/services/marker_interpolator.dart
import 'package:flutter/animation.dart';
import 'package:latlong2/latlong.dart';

/// Function signature for interpolation frame updates
typedef OnInterpolationUpdate = void Function(
  String memberId,
  LatLng interpolatedPosition,
  double interpolatedHeading,
);

/// Smooth coordinate and rotation interpolator between incoming GPS pings.
/// Eliminates marker jitter and teleports by driving high-frequency tween animations.
class MarkerInterpolator {
  final TickerProvider vsync;
  final OnInterpolationUpdate onUpdate;

  // Active animation controllers per member
  final Map<String, AnimationController> _controllers = {};
  final Map<String, LatLng> _currentPositions = {};
  final Map<String, double> _currentHeadings = {};

  MarkerInterpolator({
    required this.vsync,
    required this.onUpdate,
  });

  /// Ingests a new raw GPS ping for a member and initiates a smooth tween animation
  void updateTarget({
    required String memberId,
    required LatLng newPosition,
    required double newHeading,
    Duration duration = const Duration(milliseconds: 1500),
  }) {
    final startPosition = _currentPositions[memberId] ?? newPosition;
    final startHeading = _currentHeadings[memberId] ?? newHeading;

    // If first ping or distance is negligible, set directly without animation
    if (!_currentPositions.containsKey(memberId)) {
      _currentPositions[memberId] = newPosition;
      _currentHeadings[memberId] = newHeading;
      onUpdate(memberId, newPosition, newHeading);
      return;
    }

    // Stop existing running animation for this member
    _controllers[memberId]?.stop();
    _controllers[memberId]?.dispose();

    final controller = AnimationController(
      vsync: vsync,
      duration: duration,
    );

    // Calculate shortest angular delta for heading
    double deltaHeading = (newHeading - startHeading) % 360.0;
    if (deltaHeading > 180.0) deltaHeading -= 360.0;
    if (deltaHeading < -180.0) deltaHeading += 360.0;
    final targetAdjustedHeading = startHeading + deltaHeading;

    final curvedAnimation = CurvedAnimation(
      parent: controller,
      curve: Curves.easeOutCubic,
    );

    controller.addListener(() {
      final t = curvedAnimation.value;

      // Latitude and Longitude linear interpolation
      final lat = _lerpDouble(startPosition.latitude, newPosition.latitude, t);
      final lng = _lerpDouble(startPosition.longitude, newPosition.longitude, t);
      final heading = (_lerpDouble(startHeading, targetAdjustedHeading, t)) % 360.0;

      final current = LatLng(lat, lng);
      _currentPositions[memberId] = current;
      _currentHeadings[memberId] = heading;

      onUpdate(memberId, current, heading);
    });

    controller.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        _controllers.remove(memberId);
        controller.dispose();
      }
    });

    _controllers[memberId] = controller;
    controller.forward();
  }

  LatLng? getCurrentPosition(String memberId) => _currentPositions[memberId];
  double? getCurrentHeading(String memberId) => _currentHeadings[memberId];

  double _lerpDouble(double a, double b, double t) {
    return a + (b - a) * t;
  }

  void dispose() {
    for (final controller in _controllers.values) {
      controller.dispose();
    }
    _controllers.clear();
  }
}

File: mobile/src/services/MarkerInterpolator.ts
export interface LatLng {
  latitude: number;
  longitude: number;
}

export type OnInterpolationCallback = (
  memberId: string,
  position: LatLng,
  heading: number
) => void;

interface ActiveAnimation {
  memberId: string;
  startPos: LatLng;
  targetPos: LatLng;
  startHeading: number;
  targetHeading: number;
  startTime: number;
  duration: number;
  rafId?: number;
}

export class MarkerInterpolator {
  private currentPositions = new Map<string, LatLng>();
  private currentHeadings = new Map<string, number>();
  private activeAnimations = new Map<string, ActiveAnimation>();
  private onUpdate: OnInterpolationCallback;

  constructor(onUpdate: OnInterpolationCallback) {
    this.onUpdate = onUpdate;
  }

  public updateTarget(params: {
    memberId: string;
    newPosition: LatLng;
    newHeading: number;
    durationMs?: number;
  }): void {
    const { memberId, newPosition, newHeading, durationMs = 1200 } = params;

    const existingPos = this.currentPositions.get(memberId);
    const existingHeading = this.currentHeadings.get(memberId) ?? newHeading;

    // First fix: set immediately without animation
    if (!existingPos) {
      this.currentPositions.set(memberId, newPosition);
      this.currentHeadings.set(memberId, newHeading);
      this.onUpdate(memberId, newPosition, newHeading);
      return;
    }

    // Cancel existing animation for this member
    const existingAnim = this.activeAnimations.get(memberId);
    if (existingAnim && existingAnim.rafId) {
      cancelAnimationFrame(existingAnim.rafId);
      this.activeAnimations.delete(memberId);
    }

    // Calculate shortest angular delta
    let deltaHeading = (newHeading - existingHeading) % 360;
    if (deltaHeading > 180) deltaHeading -= 360;
    if (deltaHeading < -180) deltaHeading += 360;
    const targetAdjustedHeading = existingHeading + deltaHeading;

    const anim: ActiveAnimation = {
      memberId,
      startPos: existingPos,
      targetPos: newPosition,
      startHeading: existingHeading,
      targetHeading: targetAdjustedHeading,
      startTime: Date.now(),
      duration: durationMs,
    };

    const step = () => {
      const elapsed = Date.now() - anim.startTime;
      const progress = Math.min(elapsed / anim.duration, 1.0);

      // Ease out cubic: 1 - pow(1 - x, 3)
      const t = 1 - Math.pow(1 - progress, 3);

      const lat = anim.startPos.latitude + (anim.targetPos.latitude - anim.startPos.latitude) * t;
      const lng = anim.startPos.longitude + (anim.targetPos.longitude - anim.startPos.longitude) * t;
      const heading = (anim.startHeading + (anim.targetHeading - anim.startHeading) * t + 360) % 360;

      const current = { latitude: lat, longitude: lng };
      this.currentPositions.set(memberId, current);
      this.currentHeadings.set(memberId, heading);

      this.onUpdate(memberId, current, heading);

      if (progress < 1.0) {
        anim.rafId = requestAnimationFrame(step);
      } else {
        this.activeAnimations.delete(memberId);
      }
    };

    this.activeAnimations.set(memberId, anim);
    anim.rafId = requestAnimationFrame(step);
  }

  public getCurrentPosition(memberId: string): LatLng | undefined {
    return this.currentPositions.get(memberId);
  }

  public getCurrentHeading(memberId: string): number | undefined {
    return this.currentHeadings.get(memberId);
  }

  public dispose(): void {
    for (const anim of this.activeAnimations.values()) {
      if (anim.rafId) cancelAnimationFrame(anim.rafId);
    }
    this.activeAnimations.clear();
    this.currentPositions.clear();
    this.currentHeadings.clear();
  }
}
